File: cpp/equity-var-engine/src/historical.cpp
#include "eqvar/historical.hpp"

#include <algorithm>
#include <cmath>
#include <numeric>
#include <stdexcept>
#include <string>

#include "eqvar/returns.hpp"

namespace eqvar {

double quantile_linear(std::span<const double> x, double q) {
    if (x.empty()) throw std::invalid_argument("quantile_linear: empty sample");
    if (!(q >= 0.0 && q <= 1.0)) {
        throw std::invalid_argument("quantile_linear: q must be in [0, 1], got " +
                                    std::to_string(q));
    }
    std::vector<double> sorted(x.begin(), x.end());
    std::sort(sorted.begin(), sorted.end());
    // NumPy default "linear" (Hyndman-Fan type 7) interpolation.
    const double h = q * static_cast<double>(sorted.size() - 1);
    const std::size_t lo = static_cast<std::size_t>(std::floor(h));
    if (lo + 1 >= sorted.size()) return sorted.back();
    const double frac = h - static_cast<double>(lo);
    return sorted[lo] + frac * (sorted[lo + 1] - sorted[lo]);
}

double historical_var(std::span<const double> pnl, double alpha) {
    validate_alpha(alpha);
    validate_pnl(pnl, kMinHistObs);
    return -quantile_linear(pnl, alpha);
}

std::vector<double> brw_weights(std::size_t n, double lam) {
    if (!(lam > 0.0 && lam < 1.0)) {
        throw std::invalid_argument("brw_weights: decay lam must be in (0, 1), got " +
                                    std::to_string(lam));
    }
    if (n < 1) throw std::invalid_argument("brw_weights: n must be >= 1");
    // w_i = (1 - lam) lam^{age} / (1 - lam^n), age = n-1-i (0 = oldest).
    std::vector<double> w(n);
    const double norm = (1.0 - lam) / (1.0 - std::pow(lam, static_cast<double>(n)));
    for (std::size_t i = 0; i < n; ++i) {
        w[i] = norm * std::pow(lam, static_cast<double>(n - 1 - i));
    }
    return w;
}

double age_weighted_var(std::span<const double> pnl, double alpha, double lam) {
    validate_alpha(alpha);
    validate_pnl(pnl, kMinHistObs);
    const std::vector<double> w = brw_weights(pnl.size(), lam);
    // Stable argsort by P&L ascending, then invert the weighted step CDF.
    std::vector<std::size_t> order(pnl.size());
    std::iota(order.begin(), order.end(), std::size_t{0});
    std::stable_sort(order.begin(), order.end(),
                     [&pnl](std::size_t a, std::size_t b) { return pnl[a] < pnl[b]; });
    double cum = 0.0;
    for (std::size_t r = 0; r < order.size(); ++r) {
        cum += w[order[r]];
        if (cum >= alpha) return -pnl[order[r]];  // searchsorted-left semantics
    }
    return -pnl[order.back()];
}

std::vector<double> ewma_volatility(std::span<const double> x, double lam) {
    if (!(lam > 0.0 && lam < 1.0)) {
        throw std::invalid_argument("ewma_volatility: decay lam must be in (0, 1), got " +
                                    std::to_string(lam));
    }
    if (x.size() < 2) {
        throw std::invalid_argument("ewma_volatility: need at least 2 observations");
    }
    // Seed with the population (ddof = 0) sample variance, matching the
    // Python reference (np.var default).
    double mu = 0.0;
    for (double v : x) mu += v;
    mu /= static_cast<double>(x.size());
    double seed = 0.0;
    for (double v : x) seed += (v - mu) * (v - mu);
    seed /= static_cast<double>(x.size());
    if (seed <= 0.0) seed = 1e-16;  // zero-variance series: avoid division by 0

    std::vector<double> sigma(x.size());
    double sig2 = seed;
    sigma[0] = std::sqrt(std::max(sig2, 1e-32));
    for (std::size_t t = 1; t < x.size(); ++t) {
        sig2 = lam * sig2 + (1.0 - lam) * x[t - 1] * x[t - 1];
        sigma[t] = std::sqrt(std::max(sig2, 1e-32));
    }
    return sigma;
}

double filtered_historical_var(std::span<const double> pnl, double alpha, double lam) {
    validate_alpha(alpha);
    validate_pnl(pnl, kMinHistObs);
    const std::vector<double> sigma = ewma_volatility(pnl, lam);
    const double last = pnl[pnl.size() - 1];
    const double sig_last = sigma.back();
    const double sigma_next = std::sqrt(lam * sig_last * sig_last + (1.0 - lam) * last * last);
    std::vector<double> scenarios(pnl.size());
    for (std::size_t t = 0; t < pnl.size(); ++t) scenarios[t] = pnl[t] / sigma[t] * sigma_next;
    return -quantile_linear(scenarios, alpha);
}

double scale_var_sqrt_time(double var_1d, int horizon_days) {
    if (horizon_days < 1) {
        throw std::invalid_argument("scale_var_sqrt_time: horizon_days must be >= 1, got " +
                                    std::to_string(horizon_days));
    }
    return var_1d * std::sqrt(static_cast<double>(horizon_days));
}

}  // namespace eqvar

File: rust/equity-var-engine/src/historical.rs
//! Historical-simulation VaR: plain, age-weighted (BRW) and filtered (FHS).
//!
//! Conventions (identical to the Python reference `eq_var.historical_var`):
//!
//! * `pnl` slices are P&L in currency units, **loss < 0**;
//! * `alpha` is the tail probability (`0.01` → 99 % VaR);
//! * VaR is reported as a **positive** number: `VaR_alpha = -Q_alpha(pnl)`;
//! * plain historical VaR uses NumPy's default `"linear"` (Hyndman–Fan
//!   type-7) interpolation between order statistics — the most common desk
//!   choice; alternatives (lower/higher order statistic) differ by `O(1/n)`
//!   and are discussed in the Python project's docs/METHODOLOGY.md;
//! * weighted quantiles (BRW) use the step-function inversion of the
//!   weighted empirical CDF.

use crate::{validate_alpha, EqVarError, Result};

/// Minimum P&L observations for a historical tail estimate: fewer than this
/// cannot resolve a 1–5 % tail sensibly.
pub const MIN_OBS: usize = 50;

fn validate_pnl(pnl: &[f64], min_obs: usize) -> Result<()> {
    if pnl.len() < min_obs {
        return Err(EqVarError::InvalidInput(format!(
            "need at least {min_obs} P&L observations for historical VaR, got {}; \
             empirical tail quantiles are meaningless on shorter samples",
            pnl.len()
        )));
    }
    if pnl.iter().any(|v| !v.is_finite()) {
        return Err(EqVarError::InvalidInput(
            "pnl contains NaN or infinite values".to_string(),
        ));
    }
    Ok(())
}

fn sorted_copy(pnl: &[f64]) -> Vec<f64> {
    let mut v = pnl.to_vec();
    v.sort_by(|a, b| a.partial_cmp(b).expect("finite values compare totally"));
    v
}

/// Empirical quantile with NumPy-default linear (Hyndman–Fan type-7)
/// interpolation: `h = (n - 1) q`, result
/// `x_(floor(h)) + (h - floor(h)) * (x_(floor(h)+1) - x_(floor(h)))` on the
/// ascending order statistics.
///
/// `q` must be in `[0, 1]`; input needs at least one finite observation.
pub fn linear_quantile(pnl: &[f64], q: f64) -> Result<f64> {
    if pnl.is_empty() {
        return Err(EqVarError::InvalidInput(
            "quantile of an empty sample is undefined".to_string(),
        ));
    }
    if !(0.0..=1.0).contains(&q) {
        return Err(EqVarError::InvalidInput(format!(
            "quantile level must be in [0, 1], got {q}"
        )));
    }
    if pnl.iter().any(|v| !v.is_finite()) {
        return Err(EqVarError::InvalidInput(
            "pnl contains NaN or infinite values".to_string(),
        ));
    }
    let sorted = sorted_copy(pnl);
    Ok(linear_quantile_sorted(&sorted, q))
}

/// Type-7 quantile on an already ascending-sorted, non-empty slice.
pub(crate) fn linear_quantile_sorted(sorted: &[f64], q: f64) -> f64 {
    let n = sorted.len();
    if n == 1 {
        return sorted[0];
    }
    let h = (n - 1) as f64 * q;
    let lo = h.floor() as usize;
    if lo + 1 >= n {
        return sorted[n - 1];
    }
    let frac = h - lo as f64;
    sorted[lo] + frac * (sorted[lo + 1] - sorted[lo])
}

/// Plain historical-simulation VaR (equal weights).
///
/// Returns `-quantile_alpha(pnl)` (positive for a loss) with linear
/// (type-7) interpolation between order statistics. Requires at least
/// [`MIN_OBS`] observations and `alpha` in `(0, 0.5)`.
///
/// # Examples
///
/// ```
/// use eq_var_engine::historical::historical_var;
/// let pnl: Vec<f64> = (1..=100).map(|i| -(i as f64)).collect();
/// let var = historical_var(&pnl, 0.05).unwrap();
/// assert!((var - 95.05).abs() < 1e-12); // hand-computed type-7 quantile
/// ```
pub fn historical_var(pnl: &[f64], alpha: f64) -> Result<f64> {
    validate_alpha(alpha)?;
    validate_pnl(pnl, MIN_OBS)?;
    Ok(-linear_quantile(pnl, alpha)?)
}

/// Boudoukh–Richardson–Whitelaw exponential age weights.
///
/// Observation `i` (0 = oldest, `n - 1` = most recent) gets weight
/// proportional to `lam^{n-1-i}`; weights sum to 1 exactly:
/// `w_i = (1 - lam) lam^{n-1-i} / (1 - lam^n)`.
pub fn brw_weights(n: usize, lam: f64) -> Result<Vec<f64>> {
    if !(lam > 0.0 && lam < 1.0) {
        return Err(EqVarError::InvalidInput(format!(
            "decay lam must be in (0, 1), got {lam}"
        )));
    }
    if n < 1 {
        return Err(EqVarError::InvalidInput("n must be >= 1".to_string()));
    }
    let norm = (1.0 - lam) / (1.0 - lam.powi(n as i32));
    Ok((0..n)
        .map(|i| norm * lam.powi((n - 1 - i) as i32))
        .collect())
}

/// Age-weighted (BRW) historical VaR.
///
/// Recent observations receive exponentially larger weight
/// (`w_t ~ lam^age`). The weighted empirical CDF is inverted at `alpha`:
/// VaR is minus the smallest P&L whose cumulative weight (ascending P&L
/// order, stable sort) reaches `alpha`.
///
/// `lam -> 1` recovers plain historical simulation up to the
/// interpolation-scheme difference (BRW uses the step-CDF inversion).
pub fn age_weighted_var(pnl: &[f64], alpha: f64, lam: f64) -> Result<f64> {
    validate_alpha(alpha)?;
    validate_pnl(pnl, MIN_OBS)?;
    let w = brw_weights(pnl.len(), lam)?;
    // Stable argsort by P&L ascending (ties keep chronological order).
    let mut order: Vec<usize> = (0..pnl.len()).collect();
    order.sort_by(|&a, &b| pnl[a].partial_cmp(&pnl[b]).expect("finite values"));
    let mut cum = 0.0;
    let mut idx = pnl.len() - 1;
    for (rank, &i) in order.iter().enumerate() {
        cum += w[i];
        if cum >= alpha {
            idx = rank;
            break;
        }
    }
    Ok(-pnl[order[idx]])
}

/// One-step-ahead EWMA (RiskMetrics) volatility forecasts.
///
/// `sigma2[t] = lam * sigma2[t-1] + (1 - lam) * x[t-1]^2` — `sigma[t]` is
/// the forecast for day `t` made with information up to `t - 1`, so
/// standardising `x[t] / sigma[t]` uses no look-ahead. Seeded with the
/// full-sample population variance (`ddof = 0`, matching the Python
/// reference default `init="sample"`); a zero-variance series is floored at
/// 1e-16 to avoid downstream division by zero. Returns one `sigma[t] > 0`
/// per input observation.
pub fn ewma_volatility(x: &[f64], lam: f64) -> Result<Vec<f64>> {
    if !(lam > 0.0 && lam < 1.0) {
        return Err(EqVarError::InvalidInput(format!(
            "decay lam must be in (0, 1), got {lam}"
        )));
    }
    if x.len() < 2 {
        return Err(EqVarError::InvalidInput(
            "need at least 2 observations for EWMA volatility".to_string(),
        ));
    }
    let mut seed = crate::stats::population_variance(x)?;
    if seed <= 0.0 {
        seed = 1e-16;
    }
    let mut sig2 = vec![0.0; x.len()];
    sig2[0] = seed;
    for t in 1..x.len() {
        sig2[t] = lam * sig2[t - 1] + (1.0 - lam) * x[t - 1] * x[t - 1];
    }
    Ok(sig2.iter().map(|v| v.max(1e-32).sqrt()).collect())
}

/// Filtered historical simulation (FHS) VaR — the industry workhorse.
///
/// Barone-Adesi et al. / Hull–White devolatilisation:
///
/// 1. one-step-ahead EWMA vol forecasts `sigma_t` for each day;
/// 2. standardise `z_t = pnl_t / sigma_t` (i.i.d.-ish innovations);
/// 3. rescale every innovation to tomorrow's forecast
///    `sigma_{T+1}^2 = lam * sigma_T^2 + (1 - lam) * pnl_T^2`;
/// 4. empirical `alpha` quantile (type-7) of the rescaled scenarios.
///
/// VaR responds to the current vol regime while keeping the empirical
/// (fat-tailed, skewed) shape of the standardised innovations.
pub fn filtered_historical_var(pnl: &[f64], alpha: f64, lam: f64) -> Result<f64> {
    validate_alpha(alpha)?;
    validate_pnl(pnl, MIN_OBS)?;
    let sigma = ewma_volatility(pnl, lam)?;
    let n = pnl.len();
    let sigma_next =
        (lam * sigma[n - 1] * sigma[n - 1] + (1.0 - lam) * pnl[n - 1] * pnl[n - 1]).sqrt();
    let scenarios: Vec<f64> = pnl
        .iter()
        .zip(sigma.iter())
        .map(|(p, s)| p / s * sigma_next)
        .collect();
    Ok(-linear_quantile(&scenarios, alpha)?)
}

/// Square-root-of-time scaling: `VaR_h = VaR_1 * sqrt(h)`.
///
/// Valid only for i.i.d. returns with zero drift; understates multi-day
/// risk under volatility clustering / autocorrelation (docs/VALIDATION.md).
pub fn scale_var_sqrt_time(var_1d: f64, horizon_days: u32) -> Result<f64> {
    if horizon_days < 1 {
        return Err(EqVarError::InvalidInput(format!(
            "horizon_days must be >= 1, got {horizon_days}"
        )));
    }
    Ok(var_1d * (horizon_days as f64).sqrt())
}

/// Overlapping h-day P&L sums for direct multi-day historical VaR.
///
/// Returns the rolling sums `sum(pnl[t..t+h])` for `t = 0..=n-h`. Caveat:
/// overlapping sums are serially dependent, so the effective sample size is
/// ~`n/h` and quantile standard errors are much larger than the nominal
/// count suggests.
pub fn overlapping_horizon_pnl(pnl: &[f64], horizon_days: usize) -> Result<Vec<f64>> {
    if horizon_days < 1 {
        return Err(EqVarError::InvalidInput(format!(
            "horizon_days must be >= 1, got {horizon_days}"
        )));
    }
    if pnl.len() < horizon_days + MIN_OBS {
        return Err(EqVarError::InvalidInput(format!(
            "need at least {} observations for {horizon_days}-day overlapping windows, got {}",
            horizon_days + MIN_OBS,
            pnl.len()
        )));
    }
    let mut csum = vec![0.0; pnl.len() + 1];
    for (i, v) in pnl.iter().enumerate() {
        csum[i + 1] = csum[i] + v;
    }
    Ok((0..=pnl.len() - horizon_days)
        .map(|t| csum[t + horizon_days] - csum[t])
        .collect())
}
